server/aws-lsp-codewhisperer/src/language-server/telemetryService.ts

```diff
@@ -52,7 +54,24 @@ export class TelemetryService extends CodeWhispererServiceToken {
         return suggestionState
     }
 
+    private isLoginInvalidForTelemetry(): boolean {
+        return this.loginType === 'iam' || (this.loginType !== 'identityCenter' && this.optOutPreference !== 'OPTOUT')
+    }
+
+    private invokeSendTelemetryEvent(event: any) {
+        this.sendTelemetryEvent({
+            telemetryEvent: {
+                userTriggerDecisionEvent: event,
+            },
+            userContext: this.userContext,
+            optOutPreference: this.optOutPreference,
+        })
+    }
+
     public emitUserTriggerDecision(session: CodeWhispererSession) {
+        if (this.isLoginInvalidForTelemetry()) {
+            return
+        }
         const completionSessionResult = session.completionSessionResult ?? {}
         const acceptedItemId = Object.keys(completionSessionResult).find(k => completionSessionResult[k].accepted)
```

Comment:
you can simplify finding accepted item: ID is stored in `Session.acceptedSuggestionId` value